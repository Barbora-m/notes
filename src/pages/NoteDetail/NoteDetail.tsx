import { useCallback, useState } from "react";
import {
  Button,
  Col,
  Container,
  Input,
  InputGroup,
  Row,
  Spinner,
} from "reactstrap";
import { useTranslation } from "react-i18next";
import { useParams, Link, useHistory } from "react-router-dom";
import styled from "styled-components";

import { client } from "../../client";
import { useFetch } from "../../hooks";
import { Note } from "../../types";

export const NoteDetail = () => {
  const { push } = useHistory();
  const { id } = useParams<{ id: string }>();
  const { t } = useTranslation();

  const [newContent, setNewContent] = useState("");
  const [isEditing, setIsEditing] = useState(false);

  const { response: note, isLoading } = useFetch<Note>({
    url: `/${id}`,
    method: "GET",
  });

  const handleContentChange = useCallback((event) => {
    setNewContent(event.target.value);
  }, []);

  const handleSaveClick = useCallback(() => {
    try {
      void client.put(`/${id}`, { content: newContent });

      setIsEditing(false);
    } catch (error) {
      console.log(error.message);
    }
  }, [id, newContent]);

  const handleCancelClick = useCallback(() => {
    setIsEditing(false);
  }, []);

  const handleContentClick = useCallback(() => {
    setIsEditing(true);
    setNewContent(note?.content || "");
  }, [note]);

  const handleDeleteClick = useCallback(() => {
    try {
      void client.delete(`/${id}`);

      push("/notes");
    } catch (error) {
      console.log(error.message);
    }
  }, [id, push]);

  return (
    <Container>
      <Row>
        <Col>
          <Title>
            {t("noteDetailTitle")} {id}
          </Title>
        </Col>
      </Row>
      <Row>
        <Col>
          {isEditing ? (
            <InputGroup>
              <ContentInput
                autoFocus
                onChange={handleContentChange}
                value={newContent}
              />
              <SaveButton
                color="success"
                disabled={newContent === note?.content}
                onClick={handleSaveClick}
                size="sm"
              >
                {t("save")}
              </SaveButton>
              <CancelButton onClick={handleCancelClick} size="sm">
                {t("cancel")}
              </CancelButton>
            </InputGroup>
          ) : (
            <ContentWrapper onClick={handleContentClick}>
              {isLoading ? (
                <Spinner size="sm" color="success" />
              ) : (
                note?.content
              )}
            </ContentWrapper>
          )}
        </Col>
      </Row>
      <Row>
        <ButtonsCol>
          <Link to="/notes">
            <Button size="sm">{t("back")}</Button>
          </Link>
          <Button color="danger" onClick={handleDeleteClick} size="sm">
            {t("delete")}
          </Button>
        </ButtonsCol>
      </Row>
    </Container>
  );
};

const Title = styled.h5`
  margin: 15px 0;
`;

const ContentInput = styled(Input)`
  &:focus {
    box-shadow: none;
  }
`;

const SaveButton = styled(Button)`
  border-radius: 0;
`;

const CancelButton = styled(Button)`
  border-bottom-left-radius: 0;
  border-top-left-radius: 0;
`;

const ContentWrapper = styled.div`
  border-radius: 0.25rem;
  border: 1px solid #ced4da;
  color: #495057;
  cursor: pointer;
  font-size: 1rem;
  font-weight: 400;
  height: calc(1.5em + 0.75rem + 2px);
  line-height: 1.5;
  padding: 0.375rem 0.75rem;
`;

const ButtonsCol = styled(Col)`
  display: flex;
  justify-content: space-between;
  margin-top: 20px;
  padding-bottom: 10px;
`;
