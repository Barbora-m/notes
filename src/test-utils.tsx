import { ReactElement } from "react";
import { I18nextProvider } from "react-i18next";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";

import { i18n } from "./i18n";

export const renderWithProviders = (component: ReactElement) =>
  renderer.create(
    <I18nextProvider i18n={i18n}>
      <MemoryRouter>{component}</MemoryRouter>
    </I18nextProvider>
  );
