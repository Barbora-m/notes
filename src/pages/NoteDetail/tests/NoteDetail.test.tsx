import { renderWithProviders } from "../../../test-utils";
import { NoteDetail } from "../NoteDetail";

test("NoteDetail", () => {
  expect(renderWithProviders(<NoteDetail />)).toMatchSnapshot();
});
