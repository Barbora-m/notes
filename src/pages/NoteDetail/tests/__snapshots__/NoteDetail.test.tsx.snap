// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteDetail 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h5
        className="sc-bdfBwQ ceVbUX"
      >
        Note
         
      </h5>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="sc-eCssSg gIItej"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="sc-jSgupP fipNLF col"
    >
      <a
        href="/notes"
        onClick={[Function]}
      >
        <button
          aria-label={null}
          className="btn btn-secondary btn-sm"
          onClick={[Function]}
        >
          Back
        </button>
      </a>
      <button
        aria-label={null}
        className="btn btn-danger btn-sm"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
